libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m

```diff
@@ -122,84 +126,83 @@ - (void)setCustomRequestBodyStream:(NSInputStream* (^)(void))bodyInputStreamBloc
+            self.path = [self.path substringFromIndex:self.endpoint.length];
         }
-
-        // Adds query parameters to the request if any are set.
-        if (self.queryParams) {
-
-            // Not using NSUrlQueryItems because of https://stackoverflow.com/questions/41273994/special-characters-not-being-encoded-properly-inside-urlqueryitems
-            [fullUrl appendString:[SFRestRequest toQueryString:self.queryParams]];
+    }
+    
+    // Puts the pieces together and constructs a full URL.
+    fullUrl = [[NSMutableString alloc] initWithString:baseUrl];
+    if (![fullUrl hasSuffix:@"/"]) {
+        [fullUrl appendString:@"/"];
+    }
+    
+    // 'endpoint' could be empty for a custom endpoint like 'apexrest'.
+    NSMutableString *endpoint = [[NSMutableString alloc] initWithString:self.endpoint];
+    if (endpoint.length > 0) {
+        if ([endpoint hasPrefix:@"/"]) {
+            [endpoint deleteCharactersInRange:NSMakeRange(0, 1)];
+        }
+        if (![endpoint hasSuffix:@"/"]) {
+            [endpoint appendString:@"/"];
         }
-        self.request = [[NSMutableURLRequest alloc] initWithURL:[[NSURL alloc] initWithString:fullUrl]];
+        [fullUrl appendString:endpoint];
+    }
+    NSMutableString *path = [[NSMutableString alloc] initWithString:self.path];
+    if ([path hasPrefix:@"/"]) {
+        [path deleteCharactersInRange:NSMakeRange(0, 1)];
+    }
+    [fullUrl appendString:path];
 
-        // Sets HTTP method on the request.
-        [self.request setHTTPMethod:[SFRestRequest httpMethodFromSFRestMethod:self.method]];
+    // Adds query parameters to the request if any are set.
+    if (self.queryParams) {
 
-        // Sets OAuth Bearer token header on the request (if not already present).
-        if (self.requiresAuthentication && ![self.request.allHTTPHeaderFields.allKeys containsObject:@"Authorization"]) {
-            NSString *bearer = [NSString stringWithFormat:@"Bearer %@", user.credentials.accessToken];
-            [self.request setValue:bearer forHTTPHeaderField:@"Authorization"];
-        }
+        // Not using NSUrlQueryItems because of https://stackoverflow.com/questions/41273994/special-characters-not-being-encoded-properly-inside-urlqueryitems
+        [fullUrl appendString:[SFRestRequest toQueryString:self.queryParams]];
+    }
+    self.request = [[NSMutableURLRequest alloc] initWithURL:[[NSURL alloc] initWithString:fullUrl]];
+
+    //Set the service host type
+    NSURLRequestNetworkServiceType serviceType = [self urlRequestServiceType:self.networkServiceType];
+    [self.request setNetworkServiceType:serviceType];
+    
+    // Sets HTTP method on the request.
+    [self.request setHTTPMethod:[SFRestRequest httpMethodFromSFRestMethod:self.method]];
+
+    // Sets OAuth Bearer token header on the request (if not already present).
+    if (self.requiresAuthentication && ![self.request.allHTTPHeaderFields.allKeys containsObject:@"Authorization"]) {
```

Comment:
This should be:
```
if (user && self.requiresAuthentication && ![self.request.allHTTPHeaderFields.allKeys containsObject:@"Authorization"])
```
This guards against someone who has an unauthenticated instance (`user == nil`), but has `self.requiresAuthentication == YES`.